src/site/antora/modules/ROOT/pages/faq.adoc

```diff
@@ -17,404 +17,196 @@ Licensed to the Apache Software Foundation (ASF) under one or more
 
 [#reconfig_from_code]
-== How do I reconfigure log4j2 in code with a specific configuration file?
+== How do I reconfigure Log4j programmatically?
 
-See the below example.
-Be aware that this LoggerContext class is not part of the public API so your code may break with any minor release.
+You can reconfigure Log4j programmatically as follows:
 
 [source,java]
 ----
-// import org.apache.logging.log4j.core.LoggerContext;
-
-LoggerContext context = (org.apache.logging.log4j.core.LoggerContext) LogManager.getContext(false);
-File file = new File("path/to/a/different/log4j2.xml");
+LoggerContext context = (org.apache.logging.log4j.core.LoggerContext) LogManager.getContext(false); // <1>
+File file = new File("/path/to/a/different/log4j2.xml");
 
-// this will force a reconfiguration
+// This will force a reconfiguration
 context.setConfigLocation(file.toURI());
 ----
+<1> Be aware that this `LoggerContext` class is not part of the public API.
+Your code may break with any minor release.
 
 [#shutdown]
-== How do I shut down log4j2 in code?
+== How do I shut down Log4j programmatically?
 
 Normally there is no need to do this manually.
-Each `LoggerContext` registers a shutdown hook that takes care of releasing resources when the JVM exits (unless system property `log4j.shutdownHookEnabled` is set to `false`).
-Web applications should include the log4j-web module in their classpath which disables the shutdown hook but instead cleans up log4j resources when the web application is stopped.
+Each `LoggerContext` registers a shutdown hook that takes care of releasing resources when the JVM exits, unless the `log4j.shutdownHookEnabled` system property is set to `false`.
+Likewise, xref:manual/webapp.adoc[Web applications] replace the shutdown hook with their own resource management mechanisms.
+That is, they clean up necessary resources when the web application is stopped.
+However, if you need to manually shut down Log4j, you can use one of the `shutdown()` methods of link:../javadoc/log4j-api/org/apache/logging/log4j/LogManager.html#shutdown()[`LogManager`].
 
-However, if you need to manually shut down Log4j, you can do so as in the below example.
-Note that there is an optional parameter for specifying which `LoggerContext` to shut down.
+[#reconfig_level_from_code]
+== How do I set a logger level programmatically?
+
+You can set the level of a logger using link:../javadoc/log4j-core/org/apache/logging/log4j/core/config/Configurator.html[`Configurator`] from Log4j Core:
 
 [source,java]
 ----
-import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.core.config.Configurator; // <1>
 
-// ...
+// Set the level of particular logger associated with a class
+Configurator.setLevel("com.example.Foo", Level.DEBUG);
 
-LogManager.shutdown();
+// Set the level of the root logger
+Configurator.setRootLevel(Level.DEBUG);
 ----
+<1> Be aware that `Configurator` is not part of the public API.
```

Comment:
@ppkarwasz, I am really confused by our approach here. We explain users how they can use Log4j Core to do `X`, where we sort of recognize `X` as a valid use case, yet warn them that they will be using an internal API (contrast this to Java's `Unsafe`), which implies, they better should not do it. I have a feeling that we either should not officially document its usage, or remove the warning.

(There is a similar case for the `org.apache.logging.log4j.core.LoggerContext` shared a little bit above.)

Reply:
IMHO, `Configurator` **is** public API.

I would remove all the `(LoggerContext) LogManager.getContext(false)` tricks from our documentation, since they are a source of `ClassCastException`s, whereas `Configurator` logs the errors instead of throwing.

Reply:
Fixed in 16a4e08a770abf60bf8951b5e32f11a31f48ca61.